Support NumPy 2.0+ #98 -  Initial Commit for toml and example script

reyan-singh opened the pull request:
This commit includes changes in toml file to update numpy verstion to 2.0.2 and using numba latest

along with changes in performance_plot.py to use custom modal and agent for testing

EwoutH commented:
Thanks for the PR!

I notice it changes a lot of stuff that isn't related to this PR. Could you make it a clean diff? (you can open a new PR if that's easier)

adamamer20 commented:
![boltzmann_no_mesa](https://github.com/user-attachments/assets/2efd92cf-b54e-4c62-8c10-2a9323dd1701)

with numpy 2.0 pandas seems to match the performance of polars. I don't know if this translates also to memory management

reyan-singh commented:
> Thanks for the PR!
> 
> I notice it changes a lot of stuff that isn't related to this PR. That makes it difficult to review. Could you make it a clean diff? (you can open a new PR if that's easier)

I was verifying updated files as I checked-in 4 files only, but in list it's showing 2 commits, first commit contains my changes  and hold clean diff  but there was 2nd auto commit "pre-commit.ci" hook created ci service while running auto checks

EwoutH commented:
You're right, weird that this got through the CI on the main branch.

reyan-singh commented:
> ![boltzmann_no_mesa](https://private-user-images.githubusercontent.com/136176500/425648564-2efd92cf-b54e-4c62-8c10-2a9323dd1701.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI3MTgwMjgsIm5iZiI6MTc0MjcxNzcyOCwicGF0aCI6Ii8xMzYxNzY1MDAvNDI1NjQ4NTY0LTJlZmQ5MmNmLWI1NGUtNGM2Mi04YzEwLTJhOTMyM2RkMTcwMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyM1QwODE1MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZmU4NjRlOGZmYzA2YWM1NGFlMDVkODNiYzYzZjdkZTI1ZThkMjk3NWFmM2I5MWQ3MTY0MDdiYzY3MWMwN2RiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JZqLuHuyLZJ_s2Lu4EZDYfUCOGTv3vTtmzi9sT4cHNM)
> 
> with numpy 2.0 pandas seems to match the performance of polars. I don't know if this translates also to memory management

 @adamamer20  yes agree looks like both polars and pandas now performing equally as per plot, possibly as pandas is rely on numpy may with new release of numpy 2.0 improved  memory management as I read updated version's introduces improved CoW mechanism, memory leak fix etc

reyan-singh commented:
> You're right, weird that this got through the CI on the main branch.

Thanks sir, please suggest if any action required from my end

adamamer20 commented:
@reyan-singh `mesa`, `mesa-frames pandas` and `mesa-frames polars` all have the same exact performance, down to the decimal second. It also seems there has been a decrease in performance for `mesa-frames polars`. Seems fishy. Can you double check the setup is correct?

EwoutH commented:
Could you rebase your PR? We fixed the pre-commit issue in https://github.com/projectmesa/mesa-frames/pull/127

reyan-singh commented:
> Could you rebase your PR? We fixed the pre-commit issue in https://github.com/projectmesa/mesa-frames/pull/127

Sure Sir

reyan-singh commented:
> @reyan-singh `mesa`, `mesa-frames pandas` and `mesa-frames polars` all have the same exact performance, down to the decimal second. It also seems there has been a decrease in performance for `mesa-frames polars`. Seems fishy. Can you double check the setup is correct?

@adamamer20 I rolled back to the previous Numpy version 1.26 to verify the existing functionality and tried to run the DefaultAgent to verify the default performance behavior and looks like in Mesa 3.0 and higher we cannot use the DefaultAgent as per error 

> AttributeError: You are trying to set model.agents. In Mesa 3.0 and higher, this attribute is used by Mesa itself, so you cannot use it directly anymore.Please adjust your code to use a different attribute name for custom agent storage.

Is there any way to test this using DefaultAgent to make sure that the CustomAgent is not causing this performance behavior or my  setup is using mesa > 3  version and so the reason

adamamer20 commented:
@reyan-singh 
The reason is that [`mesa` 3.0 no longer supports assigning to `model.agents`](https://mesa.readthedocs.io/stable/migration_guide.html#model-agents). Upgrading this example to `mesa` 3.0 should be handled in a separate PR.  
For now, I'll pin `mesa` to version 2.4.0 in the `project.toml` for the development environment.

reyan-singh commented:
@adamamer20  / @EwoutH  Rebased and updated PR, Please review

reyan-singh commented:
@adamamer20  Quick observation, I found that both examples including boltzmann_wealth / sugarscrape_ig was still not running with Mesa 2.4.0 and was  getting a different set of errors  (hope not missing any step)  but when i tried with other Mesa versions e.g. I tried with Mesa 3.0.0a0 and was able to run the Default agent code

Does this behaviour look correct?

reyan-singh commented:
@adamamer20  I have updated the code with latest changes to fix performance plot issue showing same values for polars/pandas/mesa as straight line.

find after few tests that default agent cannot set if version >=2.4.0 so removed manual agent additon and added version verification to run defaut agents in case of version >= 2.4.0

Also adding sample performance plots generted with latest changes

reyan-singh commented:
![boltzmann_no_mesa](https://github.com/user-attachments/assets/ad99a3c6-e6c5-45c7-a85e-3e376c39555f)
![boltzmann_with_mesa](https://github.com/user-attachments/assets/db3320ee-7ef5-4007-8bbd-26800c21c21c)

adamamer20 commented on mesa_frames/concrete/pandas/agentset.py:
```diff
@@ -73,9 +73,10 @@ def step(self):
-    """WARNING: AgentSetPandas is deprecated and will be removed in the next release of mesa-frames.
-
+    """
+    WARNING: AgentSetPandas is deprecated and will be removed in the next release of mesa-frames.
```
```suggestion
    WARNING: AgentSetPandas is deprecated and will be removed in the next release of mesa-frames.
    
```

adamamer20 commented on mesa_frames/concrete/pandas/agentset.py:
```diff
@@ -73,9 +73,10 @@ def step(self):
+    """
+    WARNING: AgentSetPandas is deprecated and will be removed in the next release of mesa-frames.
     pandas-based implementation of AgentSetDF.
+
```
```suggestion

```

Adding a blank line will make the pre-commit check pass

adamamer20 commented on mesa_frames/concrete/pandas/mixin.py:
```diff
@@ -51,9 +51,11 @@ def _some_private_method(self):
+    WARNING: PandasMixin is deprecated and will be removed in the next release of mesa-frames.
     pandas-based implementation of DataFrame operations.
-    """  # noqa: D205
+
```
```suggestion

```

Adding a blank line will make the pre-commit check pass

adamamer20 commented on mesa_frames/concrete/pandas/mixin.py:
```diff
@@ -51,9 +51,11 @@ def _some_private_method(self):
 class PandasMixin(DataFrameMixin):
-    """WARNING: PandasMixin is deprecated and will be removed in the next release of mesa-frames.
+    """
+    WARNING: PandasMixin is deprecated and will be removed in the next release of mesa-frames.
```
```suggestion
    WARNING: PandasMixin is deprecated and will be removed in the next release of mesa-frames.
    
```

adamamer20 commented on pyproject.toml:
```diff
@@ -33,7 +33,8 @@ classifiers = [
 ]
 dependencies = [
-  "numpy~=1.26",
+  "numpy~=2.0.2",
+  "numba",
```
```suggestion

```

adamamer20 replied:
```suggestion
  "numpy>=2.0.2",
```

This will make sure the latest compatible version of numpy is installed.
Also numba is already in the dev dependencies so we can delete it

adamamer20 replied:
```suggestion
  "numpy>=2.0.2",
```

This will make sure the latest compatible version of numpy is installed.
Also numba is already in the dev dependencies so we can delete it

adamamer20 reviewed (changes requested):
Almost ready to be merged, just a couple of changes to be done! Tested on my setup and it works. 
Seems both pandas and mesa got a bit faster with numpy 2.0, but still not matching polars performance.

adamamer20 commented:
There seems to be a problem with `numba` dependency on python 3.10. Let me check this out. If we can't resolve it we might set python version >= 3.11.

adamamer20 commented:
Weird, uv was resolving to `numba` to 0.53.0 which is not compatible with python 3.10 instead of 0.60.0. This occurs only in ubuntu and not on windows. Set the version of `numba` to be >= 0.60 so now it works!
